packages/loopring_v3/contracts/impl/libtransactions/OwnerChangeTransaction.sol

```diff
@@ -0,0 +1,189 @@
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.6.10;
+pragma experimental ABIEncoderV2;
+
+import "../../iface/ExchangeData.sol";
+import "../../thirdparty/BytesUtil.sol";
+import "../../lib/EIP712.sol";
+import "../../lib/FloatUtil.sol";
+import "../../lib/MathUint.sol";
+import "../../lib/SignatureUtil.sol";
+
+
+/// @title OwnerChangeTransaction
+/// @author Brecht Devos - <brecht@loopring.org>
+library OwnerChangeTransaction
+{
+    using BytesUtil            for bytes;
+    using FloatUtil            for uint;
+    using MathUint             for uint;
+    using SignatureUtil        for bytes32;
+
+    // bytes4(keccak256("transferOwnership(bytes,bytes)")
+    bytes4 constant internal MAGICVALUE = 0xd1f21f4f;
+
+    bytes32 constant public OWNERCHANGE_TYPEHASH = keccak256(
+        "OwnerChange(address owner,uint24 accountID,uint16 feeTokenID,uint256 fee,address newOwner,uint32 nonce,address walletAddress,bytes32 walletDataHash,bytes walletCalldata)"
+    );
+
+    bytes32 constant public WALLET_TYPEHASH = keccak256(
+        "Wallet(address walletAddress,bytes32 walletDataHash)"
+    );
+
+    /*event ChangeOwnerConsumed(
+        address  indexed owner,
+        address          newOwner
+    );*/
+
+    struct OwnerChange
+    {
+        address owner;
+        uint24  accountID;
+        uint16  feeTokenID;
+        uint    fee;
+        address newOwner;
+        uint32  nonce;
+        bytes32 walletHash;
+    }
+
+    // Auxiliary data for each owner change
+    struct OwnerChangeAuxiliaryData
+    {
+        bytes   signatureOldOwner;
+        bytes   signatureNewOwner;
+
+        address walletAddress;
+        bytes32 walletDataHash;
+        bytes   walletCalldata;
+    }
+
+    function process(
+        ExchangeData.State storage S,
+        ExchangeData.BlockContext memory ctx,
+        bytes memory data,
+        bytes memory auxiliaryData
+        )
+        internal
+        returns (uint /*feeETH*/)
+    {
+        OwnerChange memory accountTransfer = readOwnerChange(data);
+
+        OwnerChangeAuxiliaryData memory auxData = abi.decode(auxiliaryData, (OwnerChangeAuxiliaryData));
+
+        // Calculate the tx hash
+        bytes32 txHash = EIP712.hashPacked(
+            ctx.DOMAIN_SEPARATOR,
+            keccak256(
+                abi.encode(
+                    OWNERCHANGE_TYPEHASH,
+                    accountTransfer.owner,
+                    accountTransfer.accountID,
+                    accountTransfer.feeTokenID,
+                    accountTransfer.fee,
+                    accountTransfer.newOwner,
+                    accountTransfer.nonce,
+                    auxData.walletAddress,
+                    auxData.walletDataHash,
+                    keccak256(auxData.walletCalldata)
+                )
+            )
+        );
+
+        // Verify the signature if one is provided, otherwise fall back to an approved tx
+        if (auxData.signatureNewOwner.length > 0) {
```

Comment:
The main purpose is to make sure the new owner authorizes that he is now the owner of the account. Currently this is important because the EdDSA key is not reset, and I want to avoid creating accounts for addresses that the Ethereum address is not in control of). If we set the EdDSA key to 0 after the transaction we don't need to do it.

The other much less important purpose is that it's an extra safety check that the new owner is valid/available by needing a signature.

But resetting the EdDSA may be a smart thing to do regardless. (I think there are also some benefits of the owner actually signing the walletData, so the operator needs to provide a signature from the old owner as well for using that data. It may even be necessary in some cases now for security).

Reply:
> But resetting the EdDSA may be a smart thing to do regardless.
Agreed.

It just seems wired to me that the protocol requires a signature from the new owner. Verification of the validity of the new owner can be and should be done by client software in my opinion, not the protocol.

Reply:
Yeah sure, that's okay for me.

We just can't allow the EdDSA keys to be set without the authorization of the account owner.